### DISCUSSION TITLE
Discrepancy in schema between GraphQL Explorer and api.github.com/graphql

### DISCUSSION THREAD

**Discussion by beyang** [community/community#24432] [Apps, API and Webhooks]
<p>Repro:</p>
<ol>
<li>Go to the GitHub GraphQL API Explorer. Issue the following query:</li>
</ol>
<p>Query:</p>
<pre><code>query Repositories($ids: [ID!]!) {
  nodes(ids:$ids) {
    ... on Repository {
      id
      nameWithOwner
    }
  }
}
</code></pre>
<p>Variables:</p>
<pre><code>{
  "ids": [
    "asdf1",
    "MDEwOlJlcG9zaXRvcnkxMTQwNzkxOQ=="
  ]
}
</code></pre>
<p>Verify you get a response like the following:</p>
<pre><code>{
  "data": {
    "nodes": [
      null,
      [
        [
          "id",
          "MDEwOlJlcG9zaXRvcnkxMTQwNzkxOQ=="
        ],
        [
          "nameWithOwner",
          "beyang/elastigo"
        ]
      ]
    ]
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "nodes",
        0
      ],
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "message": "Could not resolve to a node with the global id of 'asdf1'"
    }
  ]
}
</code></pre>
<p>If you open Chrome dev tools and “copy as cURL” the request, you’ll find it returns the same on the command line.</p>
<ol start="2">
<li>
<p>Now issue the same request against <a href="https://api.github.com/graphql" rel="nofollow noopener">https://api.github.com/graphql</a> with cURL:</p>
<p>curl -H “Authorization: bearer $TOKEN” -X POST --data-binary ‘{“query”:“query Repositories($ids: [ID!]!) {\n nodes(ids:$ids) {\n … on Repository {\n id\n nameWithOwner\n }\n }\n}”,“variables”:{“ids”:[“asdf1”,“MDEwOlJlcG9zaXRvcnkxMTQwNzkxOQ==”]},“operationName”:“Repositories”}’ <a href="https://api.github.com/graphql" class="inline-onebox-loading" rel="nofollow noopener">https://api.github.com/graphql</a></p>
</li>
</ol>
<p>(Replace “$TOKEN” with your personal access token.)</p>
<p>Verify you get a response like the following:</p>
<pre><code>{
 "data": {
  "nodes": [
   null,
   {
    "id": "MDEwOlJlcG9zaXRvcnkxMTQwNzkxOQ==",
    "nameWithOwner": "beyang\/elastigo"
   }
  ]
 },
 "errors": [
  {
   "type": "NOT_FOUND",
   "path": [
    "nodes",
    0
   ],
   "locations": [
    {
     "line": 2,
     "column": 3
    }
   ],
   "message": "Could not resolve to a node with the global id of 'asdf1'"
  }
 ]
}
</code></pre>
<ol start="3">
<li>Note that the schema of the “data” field in the response differs between the GraphQL Explorer and <a href="https://api.github.com/graphql" rel="nofollow noopener">https://api.github.com/graphql</a>. In the former, you get an array of field-value pair arrays. In the latter, you get an array of objects that reflect the actual schema of the Repository object. Which is the correct schema in this case? It seems like the latter should be, but are there guarantees that this schema is obeyed even if there are “not found” errors for some of the objects requested (as is this case in these requests).</li>
</ol>
<p>Note that this is a different question from <a href="https://github.community/t/graphql-v4-differences-in-responses-between-explorer-and-curl/13601">https://github.community/t5/GitHub-API-Development-and/GraphQL-v4-differences-in-responses-between-Explorer-and-Curl/m-p/14684#M279</a>. There, the response value was different, but here, the response schema is different.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>The Explorer is provided for quick experimentation and is not meant to be authoritative. If you want to avoid these kinds of mismatches in the future, I would recommend <a href="https://developer.github.com/v4/guides/using-the-explorer/#using-graphiql" rel="nofollow noopener">using GraphiQL mentioned in the GraphQL Guides</a>.</p>
<p>I hope that helps!</p>
